components/List/examples/supplemental/MDCListItemCell.h

```diff
@@ -0,0 +1,104 @@
+/*
+ Copyright 2018-present the Material Components for iOS authors. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import <UIKit/UIKit.h>
+
+#import "MDCListBaseCell.h"
+
+/**
+ A basic Material List Cell with labels and imageviews.
+ */
+@interface MDCListItemCell : MDCListBaseCell
+
+/**
+ Sets the cell width for the self-sizing cell.
+ Note: The self-sizing is only applied to the height, and the width is set using this method.
+ */
+@property (nonatomic, assign) CGFloat cellWidth;
```

Comment:
One thing I was considering was having a private `BOOL` called `hasSetCellWidth` or something like that that would be set to `YES` in the setter for `cellWidth` and set to `NO` in `prepareForReuse`. There'd be an assertion in the layout code that `hasSetCellWidth` had been set to enforce this property's proper usage. This was just an idea I was toying with. Do you have any thoughts?

Reply:
I think that seems fragile and prone to lots of frustration by client teams.  The width of the cell may not be known ahead of time as well. I typically expect rows like this to be the full width of the collection view section minus any insets. I would hope a UICollectionViewLayout would take care of that rather than the cell being responsible for determining its size.

Reply:
Let's assume the width of the cell is the width of the collection view section minus any insets, like you said. In that case, in cellForItemAtIndexPath we'd set the cellWidth of the cell, and it would use that to calculate its height. The cell's height is dependent on the width.

Reply:
The cell overrides `-systemLayoutSizeFittingSize:`. Its implementation uses the various UIView-related properties that have been set and the cell width that is specified in the `cellWidth` property to calculate the frame for every subview. After that a height is calculated considering the final subview frames. The method ultimately returns a CGSize based on this desired width and calculated height. Later, in `layoutSubviews`, subviews are laid out according to the frames calculated before.